connexion/operations/openapi.py

```diff
@@ -0,0 +1,376 @@
+
+        :param method: HTTP method
+        :type method: str
+        :param path:
+        :type path: str
+        :param operation: swagger operation object
+        :type operation: dict
+        :param resolver: Callable that maps operationID to a function
+        :param path_parameters: Parameters defined in the path level
+        :type path_parameters: list
+        :param app_security: list of security rules the application uses by default
+        :type app_security: list
+        :param components: `Components Object
+            <https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#componentsObject>`_
+        :type components: dict
+        :param validate_responses: True enables validation. Validation errors generate HTTP 500 responses.
+        :type validate_responses: bool
+        :param strict_validation: True enables validation on invalid request parameters
+        :type strict_validation: bool
+        :param randomize_endpoint: number of random characters to append to operation name
+        :type randomize_endpoint: integer
+        :param validator_map: Custom validators for the types "parameter", "body" and "response".
+        :type validator_map: dict
+        :param pythonic_params: When True CamelCase parameters are converted to snake_case and an underscore is appended
+        to any shadowed built-ins
+        :type pythonic_params: bool
+        :param uri_parser_class: class to use for uri parseing
+        :type uri_parser_class: AbstractURIParser
+        """
+        self.components = components or {}
+
+        def component_get(oas3_name):
+            return self.components.get(oas3_name, {})
+
+        # operation overrides globals
+        security_schemes = component_get('securitySchemes')
+        app_security = operation.get('security', app_security)
+        uri_parser_class = uri_parser_class or OpenAPIURIParser
+
+        super(OpenAPIOperation, self).__init__(
+            api=api,
+            method=method,
+            path=path,
+            operation=operation,
+            resolver=resolver,
+            app_security=app_security,
+            security_schemes=security_schemes,
+            validate_responses=validate_responses,
+            strict_validation=strict_validation,
+            randomize_endpoint=randomize_endpoint,
+            validator_map=validator_map,
+            pythonic_params=pythonic_params,
+            uri_parser_class=uri_parser_class
+        )
+
+        self._definitions_map = {
+            'components': {
+                'schemas': component_get('schemas'),
+                'requestBodies': component_get('requestBodies'),
+                'parameters': component_get('parameters'),
+                'securitySchemes': component_get('securitySchemes'),
+                'responses': component_get('responses'),
+                'headers': component_get('headers'),
+            }
```

Comment:
It seems to be missing the "examples", "links", and "callbacks". All those are important components to be reachable within Connexion.

Reply:
Here's a library that resolves json references lazily https://github.com/gazpachoking/jsonref
what do you think?

Reply:
Sure, I agree we should indeed skip the additional features that only apply to OAS3.

 I would vote for eager evaluation of references. But due the case of loops we need to be able to handle that with some lazy reference resolver. Maybe a custom `dict` would do. Seems like the library you mentioned here does that. Sounds good to me! 👍I was trying the code here with some complex spec I have, some fixes here: https://github.com/dtkav/connexion/commit/ea54567150cafce39b80968b030a6ef2b4964cff feel free to cherry-pick.

Reply:
I'm very confused at what/where https://github.com/dtkav/connexion/commit/ea54567150cafce39b80968b030a6ef2b4964cff actually is. It seems to be a dangling commit, and I can't pull it down locally to cherry-pick. Maybe you committed that against my branch, and then I force-pushed it away? Any ideas on how to recover it?

Reply:
(thanks for the fixes by the way)

Reply:
Oh, sorry. This is the correct place https://github.com/rafaelcaricio/connexion/commit/ea54567150cafce39b80968b030a6ef2b4964cff

It currently lives in this branch https://github.com/rafaelcaricio/connexion/commits/oas3-support

Reply:
Thanks! I've cherry-picked your changes.

Reply:
hey @rafaelcaricio , any chance you can provide a test case for raw examples?
The spec says that examples can can be either a string primitive or an object, so it looks like it needs to support both.
 https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#examples-object

Reply:
If you have an example that is json, then it makes sense to have it be an object in the spec (this way it can be validated). If not, then connexion won't validate the object (if it's xml or something), and send it to the client as is.

I don't really understand the use case of having raw json. It seems significantly worse, in that your editor can't lint it, and the validator can't validate it. What am I missing?